File: api/python/upsert-guess.py
import json
from http.server import BaseHTTPRequestHandler
import logging
import psycopg2
import os


def get_connection():
    _name = os.environ.get("POSTGRES_DATABASE")
    _user = os.environ.get("POSTGRES_USER")
    _password = os.environ.get("POSTGRES_PASSWORD")
    _host = os.environ.get("POSTGRES_HOST")

    return psycopg2.connect(
        dbname=_name,
        user=_user,
        password=_password,
        host=_host,
        port="5432",  # Default PostgreSQL port
    )


def get_player(user_id: int, game_id: int):
    with get_connection() as conn:
        with conn.cursor() as cur:
            cur.execute(
                """ SELECT *
                    FROM players
                    WHERE user_id = %(user_id)s AND game_id = %(game_id)s""",
                {"user_id": user_id, "game_id": game_id},
            )
            result = cur.fetchone()
    return {
        "id": result[0],
        "user_id": result[1],
        "game_id": result[2],
        "spotify_playlist_id": result[3],
        "score": result[4],
    }


def get_game(game_id: int):
    with get_connection() as conn:
        with conn.cursor() as cur:
            cur.execute(
                """
                SELECT * FROM games
                WHERE id = %(game_id)s
            """,
                {"game_id": game_id},
            )
            result = cur.fetchone()
    print(f"Updated game to {result}")
    return {
        "id": result[0],
        "state": result[1],
        "current_song_id": result[2],
    }


def get_song(song_id: int):
    with get_connection() as conn:
        with conn.cursor() as cur:
            cur.execute(
                """ SELECT *
                    FROM songs
                    WHERE id = %(song_id)s""",
                {"song_id": song_id},
            )
            result = cur.fetchone()
    return {
        "id": result[0],
        "spotify_track_id": result[1],
        "player_id": result[2],
    }


def guess_exists(player_id: int, current_song_id: int):
    with get_connection() as conn:
        with conn.cursor() as cur:
            # Does guess exist?
            cur.execute(
                """ SELECT id
                    FROM guesses
                    WHERE player_id = %(player_id)s AND current_song_id = %(current_song_id)s""",
                {"player_id": player_id, "current_song_id": current_song_id},
            )
            if cur.fetchone():
                return True
            else:
                return False


def insert_guess(
    player_id: int, guessed_player_id: int, is_correct: bool, current_song_id: int
):
    with get_connection() as conn:
        with conn.cursor() as cur:
            cur.execute(
                """ 
                    INSERT INTO guesses (player_id, guessed_player_id, is_correct, current_song_id)
                    VALUES (%(player_id)s, %(guessed_player_id)s, %(is_correct)s, %(current_song_id)s)
                    RETURNING *
                """,
                {
                    "player_id": player_id,
                    "guessed_player_id": guessed_player_id,
                    "is_correct": is_correct,
                    "current_song_id": current_song_id,
                },
            )
            result = cur.fetchone()
    return {
        "id": result[0],
        "player_id": result[1],
        "guessed_player_id": result[2],
        "is_correct": result[3],
        "current_song_id": result[4],
    }


def update_guess(
    player_id: int, guessed_player_id: int, is_correct: bool, current_song_id: int
):
    with get_connection() as conn:
        with conn.cursor() as cur:
            cur.execute(
                """ 
                    UPDATE guesses
                    SET guessed_player_id=%(guessed_player_id)s, is_correct=%(is_correct)s, current_song_id=%(current_song_id)s
                    WHERE player_id=%(player_id)s
                    RETURNING *
                """,
                {
                    "player_id": player_id,
                    "guessed_player_id": guessed_player_id,
                    "is_correct": is_correct,
                    "current_song_id": current_song_id,
                },
            )
            result = cur.fetchone()
    return {
        "id": result[0],
        "player_id": result[1],
        "guessed_player_id": result[2],
        "is_correct": result[3],
        "current_song_id": result[4],
    }


def check_all_players_guessed(game_id: int, current_song_id: int):
    with get_connection() as conn:
        with conn.cursor() as cur:
            cur.execute(
                """ SELECT *
                    FROM players
                    WHERE game_id = %(game_id)s""",
                {"game_id": game_id},
            )
            players = cur.fetchmany()
            cur.execute(
                """ SELECT *
                    FROM guesses
                    WHERE current_song_id = %(current_song_id)s""",
                {"current_song_id": current_song_id},
            )
            guesses = cur.fetchmany()
    return len(guesses) == len(players)


def update_game(game_id: int, state: int = "scoring"):
    with get_connection() as conn:
        with conn.cursor() as cur:
            cur.execute(
                """ 
                    UPDATE games
                    SET state=%(state)s
                    WHERE id=%(game_id)s
                    RETURNING *
                """,
                {"game_id": game_id, "state": state},
            )
            result = cur.fetchone()
    return {
        "id": result[0],
        "state": result[1],
        "current_song_id": result[2],
    }


def update_scores(game_id: int, current_song_id: int):
    with get_connection() as conn:
        with conn.cursor() as cur:
            # select ALL guesses for current song
            cur.execute(
                """ SELECT *
                    FROM guesses
                    WHERE current_song_id = %(current_song_id)s""",
                {"current_song_id": current_song_id},
            )
            results = cur.fetchmany()
            # Calculate scores
            for result in results:
                guess = {
                    "id": result[0],
                    "player_id": result[1],
                    "guessed_player_id": result[2],
                    "is_correct": result[3],
                    "current_song_id": result[4],
                }
                if bool(guess["is_correct"]):
                    # Increment score
                    cur.execute(
                        """ UPDATE players
                            SET score = score + 1
                            WHERE id = %(player_id)s""",
                        {"player_id": guess["player_id"]},
                    )


class handler(BaseHTTPRequestHandler):
    def do_PUT(self):
        try:
            # Log that the game creation process has started
            logging.info("Upserting new player...")

            # Parse the request body
            content_length = int(self.headers["Content-Length"])
            body = self.rfile.read(content_length)
            request_data = json.loads(body)

            print(request_data)

            # Extract data from request
            user_id = request_data.get("user_id")
            game_id = request_data.get("game_id")
            guessed_player_id = request_data.get("guessed_player_id")

            # Send SQL query to create a new game
            player = get_player(user_id, game_id)
            game = get_game(game_id)
            current_song = get_song(game["current_song_id"])
            if guess_exists(player["id"], game["current_song_id"]):
                # Update guess
                guess = update_guess(
                    player["id"],
                    guessed_player_id,
                    guessed_player_id == current_song["player_id"],
                    game["current_song_id"],
                )
            else:
                # Insert guess
                guess = insert_guess(
                    player["id"],
                    guessed_player_id,
                    guessed_player_id == current_song["player_id"],
                    game["current_song_id"],
                )

            print(f"Upserted guess: {guess}")
            # Log game creation success
            logging.info(f"Upserted guess with id: {guess['id']}")

            if check_all_players_guessed(game_id, game["current_song_id"]):
                # Update game state to scoring
                update_game(game_id=game_id, state="scoring")
                update_scores(game_id, game["current_song_id"])

            # Prepare the response
            self.send_response(200)
            self.send_header("Content-type", "application/json")
            self.end_headers()

            # Send the response as JSON
            self.wfile.write(json.dumps(guess).encode("utf-8"))

        except Exception as error:
            logging.error(f"Error: {str(error)}")

            # Handle errors, respond with status 500 and error message
            self.send_response(500)
            self.send_header("Content-type", "application/json")
            self.end_headers()

            error_response = json.dumps({"error": str(error)})
            self.wfile.write(error_response.encode("utf-8"))

File: api/upsert-guess.ts
//@ts-ignore
import * as database from "./database.js";
//@ts-ignore
import * as schema from "./schema.js";
import { VercelRequest, VercelResponse } from '@vercel/node';
import Pusher from "pusher";

const pusher = new Pusher({
    appId: process.env.VITE_PUSHER_APP_ID as string,
    key: process.env.VITE_PUSHER_KEY as string,
    secret: process.env.VITE_PUSHER_SECRET as string,
    cluster: process.env.VITE_PUSHER_CLUSTER as string,
    useTLS: true
});

// Receives ???. Creates new ??? record. Updates ??? record. Returns new/updated ??? record.
// Body : { user_id, game_id, guessed_player_id }
export default async function PUT(request: VercelRequest, response: VercelResponse) {
    try {
        const user_id = request.body.user_id as number;
        const game_id = request.body.game_id as number;
        const guessed_player_id = request.body.guessed_player_id as number; // Not sure where this is going to come from, may end up being some other variable.

        if (!user_id || !game_id) {
            let errmsg = "";
            if (!user_id) {
                errmsg += "User Id ";
            }
            if (!game_id) {
                errmsg += "Game Id ";
            }
            if (!guessed_player_id) {
                errmsg += "Guessed Player Id ";
            }
            errmsg += "required";
            throw new Error(errmsg);
        }

        // Get Player with User Id and Game Id
        const player = await database.findPlayerByGameAndUserIds(user_id, game_id);
        // Check if Guess is correct
        const is_correct: boolean = await database.isGuessCorrect(game_id, guessed_player_id);
        if (is_correct === undefined || is_correct === null) {
            throw new Error(`guess is invalid (is_correct: ${is_correct}) (game_id: ${game_id}, guessed_player_id: ${guessed_player_id})`)
        }

        // Get Game
        const game = await database.findGameById(game_id);

        // Create new guess with Player ID and Guessed Player ID 
        if (!game.current_song_id) {
            console.error(`Game has no current song id set ${game}`);
            throw new Error(`Something went wrong in the game logic. Tried to access the current song id before it was set.`);
        }

        const newGuessData: schema.NewGuess = { player_id: player.id, guessed_player_id: guessed_player_id, current_song_id: game.current_song_id, is_correct: is_correct };
        const newGuess = await database.upsertGuess(newGuessData);

        // Check if ALL players have guessed
        const guesses = await database.findLatestGuessesByGameId(game_id);
        const players = await database.findPlayersByGameId(game_id);
        if (guesses.length === players.length) { // ALL players have guessed
            // Calculate new scores
            guesses.forEach(async (guess: any) => {
                if (guess.is_correct) {
                    await database.incrementPlayerScore(guess.player_id);
                }
            });

            // Update game state to scoring
            const updatedGameData: schema.UpdateableGame = { id: game_id, state: "scoring" };
            const updatedGame: schema.Game = await database.updateGame(updatedGameData);

            // Broadcast game state update to clients
            pusher.trigger(`game-${game_id}`, 'game-state-updated', {
                game: updatedGame
            });
        }

        return response.status(200).json(newGuess);
    }
    catch (error) {
        console.error(error);
        return response.status(500).json({ error });
    }
}